pkg/controller/bucketaccess/bucket_access_controller_test.go

```diff
@@ -0,0 +1,584 @@
+				if in.BucketContext["ProjectID"] != projID {
+					t.Errorf("expected %s, got %s", projID, in.BucketContext["ProjectID"])
+				}
+				return &osspec.ProvisionerGrantBucketAccessResponse{
+					Principal:               principal,
+					CredentialsFileContents: credsContents,
+					CredentialsFilePath:     credsFile,
+				}, nil
+			},
+			principal:      principal,
+			serviceAccount: "",
+		},
+		{
+			name: "AzureBlob",
+			setProtocol: func(b *v1alpha1.Bucket) {
+				b.Spec.Protocol.AzureBlob = &v1alpha1.AzureProtocol{
+					StorageAccount: account,
+					ContainerName:  bucketName,
+				}
+			},
+			protocolName: v1alpha1.ProtocolNameAzure,
+			grantFunc: func(ctx context.Context, in *osspec.ProvisionerGrantBucketAccessRequest, opts ...grpc.CallOption) (*osspec.ProvisionerGrantBucketAccessResponse, error) {
+				if in.BucketName != bucketName {
+					t.Errorf("expected %s, got %s", bucketName, in.BucketName)
+				}
+				if in.Principal != principal {
+					t.Errorf("expected %s, got %s", principal, in.Principal)
+				}
+				if in.BucketContext["StorageAccount"] != account {
+					t.Errorf("expected %s, got %s", account, in.BucketContext["StorageAccount"])
+				}
+				return &osspec.ProvisionerGrantBucketAccessResponse{
+					Principal:               principal,
+					CredentialsFileContents: credsContents,
+					CredentialsFilePath:     credsFile,
+				}, nil
+			},
+			principal:      principal,
+			serviceAccount: "",
+		},
+		{
+			name: "No Principal",
+			setProtocol: func(b *v1alpha1.Bucket) {
+				b.Spec.Protocol.S3 = &v1alpha1.S3Protocol{
+					Region:           region,
+					Version:          protocolVersion,
+					SignatureVersion: sigVersion,
+					BucketName:       bucketName,
+					Endpoint:         endpoint,
+				}
+			},
+			protocolName: v1alpha1.ProtocolNameS3,
+			grantFunc: func(ctx context.Context, in *osspec.ProvisionerGrantBucketAccessRequest, opts ...grpc.CallOption) (*osspec.ProvisionerGrantBucketAccessResponse, error) {
+				return &osspec.ProvisionerGrantBucketAccessResponse{
+					Principal:               generatedPrincipal,
+					CredentialsFileContents: credsContents,
+					CredentialsFilePath:     credsFile,
+				}, nil
+			},
+			principal:      "",
+			serviceAccount: "",
+		},
+		{
+			name: "ServiceAccount exists",
+			setProtocol: func(b *v1alpha1.Bucket) {
+				b.Spec.Protocol.S3 = &v1alpha1.S3Protocol{
+					Region:           region,
+					Version:          protocolVersion,
+					SignatureVersion: sigVersion,
+					BucketName:       bucketName,
+					Endpoint:         endpoint,
+				}
+			},
+			protocolName: v1alpha1.ProtocolNameS3,
+			grantFunc: func(ctx context.Context, in *osspec.ProvisionerGrantBucketAccessRequest, opts ...grpc.CallOption) (*osspec.ProvisionerGrantBucketAccessResponse, error) {
+				return &osspec.ProvisionerGrantBucketAccessResponse{
+					Principal:               principal,
+					CredentialsFileContents: credsContents,
+					CredentialsFilePath:     credsFile,
+				}, nil
+			},
```

Comment:
What is the reason we have both SA and Principal, providing one of the two is normal usecase.

Reply:
Yes, this is a mistake in the test.  The principal shouldn't be returned here.